IdentityCore/src/throttling/cache/MSIDThrottlingCacheService.m

```diff
@@ -0,0 +1,364 @@
+
+@property (nonatomic) NSUInteger cacheSizeInt;
+@property (nonatomic) MSIDThrottlingCacheNode *head;
+@property (nonatomic) MSIDThrottlingCacheNode *tail;
+@property (nonatomic) NSMutableDictionary *container;
+@property (nonatomic) dispatch_queue_t synchronizationQueue;
+
+@end
+
+@implementation MSIDThrottlingCacheService
+
+- (NSUInteger)cacheSize
+{
+    return self.cacheSizeInt-2;
+}
+
+- (NSUInteger)numCacheRecords
+{
+    return self.container.allKeys.count-2;
+}
+
+- (instancetype)initWithThrottlingCacheSize:(NSUInteger)cacheSize
+{
+    self = [super init];
+    if (self)
+    {
+        _cacheSizeInt = cacheSize+2;
+        //create dummy head and tail
+        _head = [[MSIDThrottlingCacheNode alloc] initWithThumbprintKey:HEAD_NODE_KEY
+                                                         errorResponse:nil
+                                                          throttleType:nil
+                                                      throttleDuration:0
+                                              prevRequestThumbprintKey:nil
+                                              nextRequestThumbprintKey:TAIL_NODE_KEY];
+        
+        _tail = [[MSIDThrottlingCacheNode alloc] initWithThumbprintKey:TAIL_NODE_KEY
+                                                         errorResponse:nil
+                                                          throttleType:nil
+                                                      throttleDuration:0
+                                              prevRequestThumbprintKey:HEAD_NODE_KEY
+                                              nextRequestThumbprintKey:nil];
+        
+        //create concurrent queue and container dictionary
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.msidthrottlingcache-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+        _container = [NSMutableDictionary new];
+        
+        [self.container setObject:_head forKey:HEAD_NODE_KEY];
+        [self.container setObject:_tail forKey:TAIL_NODE_KEY];
+        
+        
+    }
+    return self;
+}
+
++ (MSIDThrottlingCacheService *)sharedInstance:(NSUInteger)cacheSize
+{
+    static MSIDThrottlingCacheService *m_service;
+    static dispatch_once_t once_token;
+    
+    dispatch_once(&once_token, ^{
+        m_service = [[MSIDThrottlingCacheService alloc] initWithThrottlingCacheSize:cacheSize];
+    });
+    
+    return m_service;
+}
+
+/* add new node to the front of LRU cache.
+if node already exists, update and move it to the front of LRU cache */
+- (void)addRequestToCache:(NSString *)thumbprintKey
+            errorResponse:(NSError *)errorResponse
+             throttleType:(NSString *)throttleType
+         throttleDuration:(NSInteger)throttleDuration
+                    error:(NSError **)error
+{
+    __block NSError *subError = nil;
```

Comment:
Validate the input params to make sure they are all valid before processing the cache.